src/XrdCeph/XrdCephBulkAioRead.cc

```diff
@@ -0,0 +1,187 @@
+   * @param object_sz         Maximum size of the ceph objects that the file is split into
+   *
+   */
+  context = ct;
+  object_size = object_sz;
+  file_name = filename;
+  log_func = logwrapper;
+}
+
+bulkAioRead::~bulkAioRead() {
+  /**
+   * Destructor. Just clears dynamically allocated memroy.
+   */
+  clear();
+}
+
+void bulkAioRead::clear() {
+  /**
+   * Clear all dynamically alocated memory
+   */
+  buffers.clear();
+  operations.clear();
+}
+
+int bulkAioRead::addRequest(size_t obj_idx, char* out_buf, size_t size, off64_t offset) {
+  /**
+   * Prepare read request for a single ceph object. Private method.
+   *
+   * Method will allocate all (well, almost, except the string for the object name)
+   * necessary objects to submit read request to ceph. To submit the requests use
+   * `submit_and_wait_for_complete` method.
+   *
+   * @param obj_idx  number of the object (starting from zero) to read
+   * @param out_buf  output buffer, where read results should be stored
+   * @param size     number of bytes to read
+   * @param offset   offset in bytes where the read should start. Note that the offset is local to the
+   *                 ceph object. I.e. if offset is 0 and object number is 1, yo'll be reading from the
+   *                 start of the second object, not from the begining of the file.
+   *
+   * @return         zero on success, negative error code on failure
+   */
+
+  try{
+    auto &op_data = operations[obj_idx];
+    //When we start using C++17, the next two lines can be merged
+    buffers.emplace_back(out_buf);
+    auto &buf = buffers.back();
+    op_data.ceph_read_op.read(offset, size, &buf.bl, &buf.rc);
+  } catch (std::bad_alloc&) {
+   log_func((char*)"Memory allocation failed while reading file %s", file_name.c_str());
+   return -ENOMEM;
+  }
+  return 0;
+}
+
+int bulkAioRead::submit_and_wait_for_complete() {
+  /**
+   * Submit previously prepared read requests and wait for their completion
+   *
+   * To prepare read requests use `read` or `addRequest` methods.
+   *
+   * @return  zero on success, negative error code on failure
+   *
+   */
+
+  std::string obj_name;
+
+  for (auto &op_data: operations) {
+    size_t obj_idx = op_data.first;
+    //16 bytes for object hex number, 1 for dot and 1 for null-terminator
+    char object_suffix[18];
+    int sp_bytes_written;
+    sp_bytes_written = snprintf(object_suffix, sizeof(object_suffix), ".%016zx", obj_idx);
+    if (sp_bytes_written >= (int) sizeof(object_suffix)) {
+      log_func((char*)"Can not fit object suffix into buffer for file %s -- too big\n", file_name.c_str());
+      return -EFBIG;
+    }
+
+    try {
+      obj_name =  file_name + std::string(object_suffix);
+    } catch (std::bad_alloc&) {
+      log_func((char*)"Can not create object string for file %s)", file_name.c_str());
+      return -ENOMEM;
+    }
+    context->aio_operate(obj_name, op_data.second.cmpl, &op_data.second.ceph_read_op, 0);
```

Comment:
This interleaving of a possibly failed string allocation and the initiation of the operation could also be problematic, again if `~CmplPtr` can't guarantee to wait for or cancel operations already initiated.

Reply:
Now `~CmplPtr` waits for completion, so there should not be any problems with this.